zookeeper/src/main/java/com/linecorp/armeria/client/endpoint/zookeeper/ZookeeperEndpointGroup.java

```diff
@@ -0,0 +1,175 @@
+     */
+    public ZookeeperEndpointGroup(String zkConnectionStr, String zNode, int sessionTimeout,
+                                  NodeValueConverter converter) {
+
+        this.zkConnectionStr = requireNonNull(zkConnectionStr, "zkConnectionStr");
+        this.znode = requireNonNull(zNode, "zNode");
+        this.nodeValueToEndpointList = requireNonNull(converter, "nodeValueToEndpointList");
+        try {
+            zooKeeper = new ZooKeeper(zkConnectionStr, sessionTimeout, event -> {
+                if (event.getState() == KeeperState.SyncConnected) {
+                    connectionLatch.countDown();
+                }
+
+            });
+            connectionLatch.await();
+            Stat stat = zooKeeper.exists(zNode, new ZkWatcher());
+            byte[] nodeData;
+            if (stat != null) {
+                nodeData = zooKeeper.getData(zNode, false, null);
+                endpointList = nodeValueToEndpointList.convert(nodeData);
+            }
+
+        } catch (Exception e) {
+            throw new EndpointGroupException("failed to connect to ZooKeeper:" + zkConnectionStr);
+        }
+    }
+
+    /**
+     * Create a Zookeeper endpoint group with a {@link DefaultNodeValueConverter}.
+     * @param zkConnectionStr  A connection string containing a comma separated list
+     *        of host:port pairs,each corresponding to a ZooKeeper server
+     * @param zNode a zNode path e.g. {@code "/groups/productionGroups"}
+     * @param sessionTimeout   Zookeeper session timeout in milliseconds
+     */
+    public ZookeeperEndpointGroup(String zkConnectionStr, String zNode, int sessionTimeout) {
+        this(zkConnectionStr, zNode, sessionTimeout, new DefaultNodeValueConverter());
+    }
+
+    @Override
+    public List<Endpoint> endpoints() {
+        return this.endpointList;
+
+    }
+
+    public class ZkWatcher implements Watcher, StatCallback {
+        @Override
+        public void process(WatchedEvent event) {
+            String path = event.getPath();
+            if (event.getType() != Event.EventType.None) {
+                if (path != null && path.equals(znode)) {
+                    // Something has changed on the node, let's find out
+                    zooKeeper.exists(znode, true, this, null);
+                }
+            }
+
+        }
+
+        @Override
+        public void processResult(int responseCodeInt, String path, Object ctx, Stat stat) {
+            boolean exists;
+            Code responseCode = get(responseCodeInt);
+            switch (responseCode) {
+                case OK:
+                    exists = true;
+                    break;
+                case NONODE:
+                    exists = false;
+                    break;
+                case SESSIONEXPIRED:
+                case NOAUTH:
+                    return;
+                default:
+                    // Retry errors
+                    zooKeeper.exists(znode, true, this, null);
```

Comment:
suppose there is a network partition, our ZooKeeper client should always try to reconnect its  server. Since this is a Async callback, it should not block others' business